ui/src/image_annotator.test.tsx

```diff
@@ -669,6 +669,339 @@ describe('ImageAnnotator.tsx', () => {
+        const { container, getByText } = render(<XImageAnnotator model={model} />)
+        await waitForLoad()
+        const canvasEl = container.querySelectorAll('canvas')[1]
+        expect(wave.args[name]).toMatchObject(items)
+        fireEvent.click(getByText('Remove all'))
+        fireEvent.wheel(canvasEl, { deltaY: 1, ctrlKey: true })
+        expect(wave.args[name]).toMatchObject([])
+      })
+    })
+
+    describe('Active shape switching', () => {
+      it('Use "b" to switch the active shape', async () => {
+        const { container } = render(<XImageAnnotator model={model} />)
+        await waitForLoad()
+        const canvasEl = container.querySelectorAll('canvas')[1]
+        fireEvent.keyDown(canvasEl, { key: 'b' })
+
+        await waitFor(() => expect(document.querySelector('[class*="is-checked"]')).toHaveAttribute('title', 'Rectangle'))
+      })
+
+      it('Cancel rectangle creation when switching the active shape', async () => {
+        const { container, getByTitle } = render(<XImageAnnotator model={model} />)
+        await waitForLoad()
+        const canvasEl = container.querySelectorAll('canvas')[1]
+        fireEvent.click(getByTitle('Rectangle'))
+        fireEvent.mouseDown(canvasEl, { clientX: 110, clientY: 110, buttons: 1 })
+        fireEvent.keyDown(canvasEl, { key: 'b' })
+        fireEvent.click(canvasEl, { clientX: 150, clientY: 150 })
+
+        expect(wave.args[name]).toMatchObject(items)
+      })
+
+      it('Cancel polygon creation when switching the active shape', async () => {
+        const { container, getByTitle } = render(<XImageAnnotator model={model} />)
+        await waitForLoad()
+        const canvasEl = container.querySelectorAll('canvas')[1]
+        fireEvent.click(getByTitle('Polygon'))
+        fireEvent.click(canvasEl, { clientX: 10, clientY: 10 })
+        fireEvent.click(canvasEl, { clientX: 20, clientY: 20 })
+        fireEvent.click(canvasEl, { clientX: 30, clientY: 30 })
+        fireEvent.keyDown(canvasEl, { key: 'b' })
+        fireEvent.click(canvasEl, { clientX: 10, clientY: 10 })
+
+        expect(wave.args[name]).toMatchObject(items)
+      })
+
+      it('Cancel the rectangle movement when switching from "select" to "rect"', async () => {
+        const { container } = render(<XImageAnnotator model={model} />)
+        await waitForLoad()
+        const canvasEl = container.querySelectorAll('canvas')[1]
+        fireEvent.click(canvasEl, { clientX: 50, clientY: 50 })
+        fireEvent.mouseDown(canvasEl, { clientX: 50, clientY: 50, buttons: 1 })
+        fireEvent.mouseMove(canvasEl, { clientX: 55, clientY: 55, buttons: 1 })
+        fireEvent.keyDown(canvasEl, { key: 'b' })
+        fireEvent.mouseMove(canvasEl, { clientX: 60, clientY: 60, buttons: 1 })
+        fireEvent.click(canvasEl, { clientX: 60, clientY: 60 })
+
+        expect(wave.args[name]).toMatchObject([{ tag: 'person', shape: { rect: { x1: 15, x2: 105, y1: 15, y2: 105 } } }, polygon])
```

Comment:
The reason why I am expecting at the end is that the `wave.args` are updated in the `onClick` handler. I think expecting at the end is fine - we know that moving has stopped when pressing `b` key. If it was not the case, the rectangle would move by 10 on both axes, not by 5. The only case I can think this test could bring a false positive result is that rectangle is not moving before pressing `b` but only after. But we can also handle it by changing `clientX` and `clientY` to let's say 70 instead of 60.

Reply:
> Let's add expect after the first mouseMove
Ok, my previous comment is irrelevant. I am setting the `wave.args` when pressing `b` as well. So I've added the expect after `keyDown` and it works.